enable initial e2e tests for s390x.

vamsikrishna-siddu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
There are no e2e tests enabled for s390x.

After this PR:
A subset of the existing e2e tests is enabled for s390x and tested against external CI cluster. All these tests which are tagged with [s390x] are tested against s390x cluster are passing.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
    enable initial e2e tests for s390x.

```

vamsikrishna-siddu commented:
/cc @brianmcarey

alicefr commented on cluster-up/cluster/external/provider.sh:
```diff
@@ -20,6 +20,7 @@ docker_tag=\${DOCKER_TAG}
 docker_prefix=\${DOCKER_PREFIX}
 manifest_docker_prefix=\${DOCKER_PREFIX}
 image_pull_policy=\${IMAGE_PULL_POLICY:-Always}
+kubectl=\${KUBECTL}
```
Any reasons for adding this?

vamsikrishna-siddu replied:
For external cluster provider The --kubectl-path here https://github.com/kubevirt/kubevirt/blob/main/hack/functests.sh#L69 is getting set as empty. The reason for this behaviour is unlike k8s providers where the kubectl binary will be scp' ed from control node https://github.com/kubevirt/kubevirt/blob/main/cluster-up/cluster/k8s-provider-common.sh#L244, the external cluster don't have this capability. So, we have to manually export the env with KUBECTL variable. So, In order to set this env variable while running test, we have to source the KUBECTL env to provider config.

I will remove this change as the kubevirt downloads the latest kubevirtci tag and put those files in the cluster-up folder.  I have raised a pr in the kubevirtCI 
https://github.com/kubevirt/kubevirtci/pull/1238

alicefr commented on tests/libvmifact/factory.go:
```diff
@@ -76,7 +76,14 @@ func NewAlpine(opts ...libvmi.Option) *kvirtv1.VirtualMachineInstance {
 	}
 	alpineOpts = append(alpineOpts, opts...)
-	return libvmi.New(alpineOpts...)
+	vmi := libvmi.New(alpineOpts...)
+
+	// Supplied with no user data, AlpimeWithTestTooling image takes more than 200s to allow login
+	if libvmi.GetCloudInitVolume(vmi) == nil {
+		withNonEmptyUserData := libvmi.WithCloudInitNoCloud(WithDummyCloudForFastBoot())
+		withNonEmptyUserData(vmi)
+	}
+	return vmi
```
Can you please split changes not related to s390x into a separate commit

vamsikrishna-siddu replied:
done

alicefr commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -51,6 +53,29 @@ var (
+	case2:
+	  build_arch is s390x, which will be translated to s390x arch
+*/
+func TranslateBuildArch() string {
+	buildArch := os.Getenv("BUILD_ARCH")
+	var arch string
+
+	if buildArch != "" {
+		archElements := strings.Split(buildArch, "-")
+		if len(archElements) == 2 {
+			arch = archElements[1]
+		} else {
+			arch = archElements[0]
+		}
+	}
+	return arch
```
Nit: this can be simplified using early return 
``` golang 
	buildArch := os.Getenv("BUILD_ARCH")
        if buildArch == {
            return ""
        }
         archElements := strings.Split(buildArch, "-")
	if len(archElements) == 2 {
		return  archElements[1]
	} else {
		return archElements[0]
	}
```

vamsikrishna-siddu replied:
done

alicefr commented on tests/vmi_lifecycle_test.go:
```diff
@@ -372,6 +377,11 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 					vmi = libvmifact.NewCirros()
+					arch := testsuite.TranslateBuildArch()
+					// Use Alpine instead of cirros as cirros is not supported on s390x.
+					if checks.IsS390X(arch) {
```
Since you are repeating this pattern multiple times, it might beneficial to create a function that constructs a tiny guest based on the arch

vamsikrishna-siddu replied:
done

vamsikrishna-siddu commented:
Hi @alicefr  i have addressed your review comments. Can you please review it once. Thanks

vamsikrishna-siddu commented:
/cc @jschintag

jschintag commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -51,6 +56,41 @@ var (
+
+// GetVMGuestByArchitecture returns the VMGuest based on the architecture.
+func GetVMGuestByArchitecture(arch string,
+	vmi *v1.VirtualMachineInstance, loginToVMI console.LoginToFunction, opts ...libvmi.Option) (*v1.VirtualMachineInstance, console.LoginToFunction) {
```
I think this function should either always be used to construct the VMI, thus enabling better arch specific options, or alternatively be renamed into something more specific like `ReplaceVMGuestOSIfS390X`

vamsikrishna-siddu replied:
done

jschintag reviewed (commented):
@vamsikrishna-siddu thank you for the PR, i have only found one thing, otherwise this looks good.

xpivarc commented on tests/libvmifact/factory.go:
```diff
@@ -76,7 +76,14 @@ func NewAlpine(opts ...libvmi.Option) *kvirtv1.VirtualMachineInstance {
-	return libvmi.New(alpineOpts...)
+	vmi := libvmi.New(alpineOpts...)
+
+	// user-data e2e tests require cloud init volume
```
What tests are we talking about here?

vamsikrishna-siddu replied:
These tests https://github.com/kubevirt/kubevirt/blob/main/tests/vmi_lifecycle_test.go#L370

lyarwood replied:
Shouldn't we explicitly add `libvmi.WithCloudInitNoCloud` there then?

xpivarc commented on tests/network/vmi_lifecycle.go:
```diff
@@ -60,6 +61,9 @@ var _ = SIGDescribe("[crit:high][arm64][vendor:cnv-qe@redhat.com][level:componen
 			DescribeTable("[Serial]VMIs shouldn't fail after the kubelet restarts", func(bridgeNetworking bool) {
 				var vmiOptions []libvmi.Option
 
+				arch := testsuite.TranslateBuildArch()
+				checks.SkipIfS390X(arch, "failing for s390x.")
```
Please don't use skips, also how does this fail?

alicefr replied:
@xpivarc do we have an example here how to create a label for the tests?

xpivarc replied:
The labels are defined [here](https://github.com/kubevirt/kubevirt/blob/main/tests/decorators/decorators.go). We should use these labels to filter what tests will run for the s390x lane. If for whatever reason we need to skip some individual test (`It`) we should not label the `Describe` but lower leafs `Context` or `It`...  to prevent the programmatic skip

xpivarc commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -91,6 +131,13 @@ func AdjustKubeVirtResource() {
 			},
 		},
 	}
+	arch := TranslateBuildArch()
+	// Disable CPUManager Featuregate for s390x as it is not supported.
+	if arch != "s390x" {
+		kv.Spec.Configuration.DeveloperConfiguration.FeatureGates = append(kv.Spec.Configuration.DeveloperConfiguration.FeatureGates,
+			virtconfig.CPUManager,
+		)
+	}
```
Does it really matter?

vamsikrishna-siddu replied:
The Kubernetes CPU Manager feature is not supported for s390x. So, our worker nodes don't have the CPU Manager label set to true. So, if we enable the feature gate for the kubevirt. When we run the e2e tests. This function https://github.com/kubevirt/kubevirt/blob/main/tests/testsuite/kubevirtresource.go#L149 is checking any available nodes with the CPU manager label and it is not able to find them and eventually failing the tests at the initial stage.

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -73,7 +73,7 @@ import (
 )
 
-var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level:component][sig-compute]VMIlifecycle", decorators.SigCompute, decorators.VMIlifecycle, func() {
+var _ = Describe("[rfe_id:273][crit:high][[arm64]][s390x][vendor:cnv-qe@redhat.com][level:component][sig-compute]VMIlifecycle", decorators.SigCompute, decorators.VMIlifecycle, func() {
```
brackets looks incorrect

vamsikrishna-siddu replied:
fixed the brackets.

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -220,6 +220,9 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 		DescribeTable("log libvirtd debug logs should be", func(vmiLabels, vmiAnnotations map[string]string, expectDebugLogs bool) {
+			arch := testsuite.TranslateBuildArch()
+
+			checks.SkipIfS390X(arch, "failing consistently for s390x.")
```
Please make a review why certain tests fail.

vamsikrishna-siddu replied:
thanks for the review @xpivarc, sure i will make a review for the test failures.

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -372,7 +377,8 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 					vmi = libvmifact.NewCirros()
-
+					arch := testsuite.TranslateBuildArch()
+					vmi, _ = testsuite.GetVMGuestByArchitecture(arch, vmi, nil)
```
imho: This won't scale, we need to tackle this somehow...
/cc @EdDev

xpivarc reviewed (commented):
/cc @EdDev

EdDev reviewed (changes requested):
Thank you for working on adding another architecture for the e2e tests.

This PR includes several changes that requires more explanation with proper reasoning.

I understand why a tag was added to some tests (so they could be filtered), but for most other changes I do not understand the reasons and the problems they solve.

Regarding the tagging, I think we should not add more of them but to use labels instead (see https://github.com/kubevirt/kubevirt/blob/2db731d15226dd8ab1ca6a72dc3d680cfc1f6cb5/tests/decorators/decorators.go#L1).

vamsikrishna-siddu commented on tests/libvmifact/factory.go:
```diff
@@ -77,6 +77,7 @@ func NewAlpine(opts ...libvmi.Option) *kvirtv1.VirtualMachineInstance {
 	}
 	alpineOpts = append(alpineOpts, opts...)
 	return libvmi.New(alpineOpts...)
+
```
done

vamsikrishna-siddu commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -51,6 +56,41 @@ var (
+		return archElements[1]
+	}
+	return archElements[0]
+
```
done

vamsikrishna-siddu commented:
Hello everyone, Thanks for the reviews.

The initial plan I thought is to Run the tests which are ran by ARM and skip the tests which are failing. Then enable the e2e tests CI job for s390x. Later we eventually enable the failing tests and s390x arch specific tests.

**Let me explain the issues I faced and why I had to do these changes.**

Since the Cirros is not supported on s390x. I am using Alpine instead of Cirros for those tests. I have introduced a new 
 function see here. https://github.com/kubevirt/kubevirt/pull/12516/files#diff-c4eb7bf2247c837b9bcddf66d9c5531c76faf68da00b6c4616503e3679ac971fR81
This function basically takes the VMI and replaces it with alpine when the architecture is s390x.

Some tests are not using either Alpine or Cirros and the tests are using the VMI's with no container disk. See here
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R238
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-ad5e41a52fa568688290426d2e073bf72031daa4e0f4f7ba16c57d2274eaa90aR72
When I try to create a Guestless and VMI with no container Disk in s390x. The VMI is moving to failed state and not coming to running state. So, I am using the Alpine for these tests and they are passing. See here https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R240
I am not able to debug why the Guestless and VMI with no container disk are not working in s390x, but they are working in x86

Since Iam doing a crossbuild-s390x on x86 the default testsuite.GOARCH does not give me the correct architecture. So, I have introduced this TranslateBuildArch fn.
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-c4eb7bf2247c837b9bcddf66d9c5531c76faf68da00b6c4616503e3679ac971fR67

**Why Iam skipping some tests?**
These tests actually related to CPU-Model. When I was raising the PR the CPU-Model support is not there. So, I skipped those tests. I am now currently working to enable those tests.
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R920

I am skipping these tests as they require both cirros and alpine
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R359

This test Iam skipping because the VMI is not getting moved to failed state if I use Alpine instead of Guestless. I am not sure how to fix it.
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R688

Some of the tests I am yet to debug them
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6R1623

I want to understand what are Guestless VMI's and the VMI's with no container disks. Any specific reason for using those in the e2e testing. This will help me fix some of the issues.

vamsikrishna-siddu commented:
> Thank you for working on adding another architecture for the e2e tests.
> 
> This PR includes several changes that requires more explanation with proper reasoning.
> 
> I understand why a tag was added to some tests (so they could be filtered), but for most other changes I do not understand the reasons and the problems they solve.
> 
> Regarding the tagging, I think we should not add more of them but to use labels instead (see
> 
> https://github.com/kubevirt/kubevirt/blob/2db731d15226dd8ab1ca6a72dc3d680cfc1f6cb5/tests/decorators/decorators.go#L1
> 
> ).

Hi @EdDev , @xpivarc  and  @alicefr I have removed the tags and added label for s390x architecture.  I have also provided the explanation for the pr changes. Can you please review it once. Thanks

dhiller commented:
/assign @dhiller

dhiller commented:
/cc

alicefr commented:
@vamsikrishna-siddu I'm wondering if there is a possibility to abstract cirros and alpine guest and select only a single time what to use. Doing the switch in all the test cases is messy and can be easily forget it for new tests.

vamsikrishna-siddu commented:
> @vamsikrishna-siddu I'm wondering if there is a possibility to abstract cirros and alpine guest and select only a single time what to use. Doing the switch in all the test cases is messy and can be easily forget it for new tests.

Hi @alicefr  I initially tried to do like that only. But the problem here is that not only Cirros there are Guestless and some places it is just no guest at all. They are also not working for s390x. So Iam replacing them with the alpine in the tests if the arch is s390x.  See here: https://github.com/kubevirt/kubevirt/blob/main/tests/vnc_test.go#L71
https://github.com/kubevirt/kubevirt/blob/main/tests/vmi_lifecycle_test.go#L232
We can still implement the function doing the abstraction like this.
```

func GetVmByGuestOS(guestOS string, opts ...libvmi.Option) (*v1.VirtualMachineInstance,
	console.LoginToFunction) {
	var vmi *v1.VirtualMachineInstance
	var loginToVMI console.LoginToFunction
	arch := TranslateBuildArch()
	if arch != "s390x" {

		if guestOS == "cirros" {
			vmi = libvmifact.NewCirros(opts...)
			loginToVMI = console.LoginToCirros
		} else if guestOS == "guestless" {
			vmi = libvmifact.NewGuestless(opts...)
		} else if guestOS == "" {
			vmi = libvmi.New(opts...)
		}

	} else {
		vmi = libvmifact.NewAlpine(opts...)
		loginToVMI = console.LoginToAlpine
	}

	return vmi, loginToVMI

}

```
is the function looking fine?

alicefr commented:
@vamsikrishna-siddu so as far as it regards the guestless function, I would move the selection directly in the `NewGuestless` function based on the arch. The name is generic enough to allow this.

alicefr commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -91,8 +130,14 @@ func AdjustKubeVirtResource() {
 		},
 	}
+	arch := TranslateBuildArch()
+	// Disable CPUManager Featuregate for s390x as it is not supported.
```
nit: you can use `DisableFeatureGate` if the arch is s390x

xpivarc replied:
How does this interfere with the test suite?

alicefr commented:
> > @vamsikrishna-siddu I'm wondering if there is a possibility to abstract cirros and alpine guest and select only a single time what to use. Doing the switch in all the test cases is messy and can be easily forget it for new tests.
> 
> Hi @alicefr I initially tried to do like that only. But the problem here is that not only Cirros there are Guestless and some places it is just no guest at all. They are also not working for s390x. So Iam replacing them with the alpine in the tests if the arch is s390x. See here: https://github.com/kubevirt/kubevirt/blob/main/tests/vnc_test.go#L71 https://github.com/kubevirt/kubevirt/blob/main/tests/vmi_lifecycle_test.go#L232 We can still implement the function doing the abstraction like this.
> 
> ```
> 
> func GetVmByGuestOS(guestOS string, opts ...libvmi.Option) (*v1.VirtualMachineInstance,
> 	console.LoginToFunction) {
> 	var vmi *v1.VirtualMachineInstance
> 	var loginToVMI console.LoginToFunction
> 	arch := TranslateBuildArch()
> 	if arch != "s390x" {
> 
> 		if guestOS == "cirros" {
> 			vmi = libvmifact.NewCirros(opts...)
> 			loginToVMI = console.LoginToCirros
> 		} else if guestOS == "guestless" {
> 			vmi = libvmifact.NewGuestless(opts...)
> 		} else if guestOS == "" {
> 			vmi = libvmi.New(opts...)
> 		}
> 
> 	} else {
> 		vmi = libvmifact.NewAlpine(opts...)
> 		loginToVMI = console.LoginToAlpine
> 	}
> 
> 	return vmi, loginToVMI
> 
> }
> ```
> 
> is the function looking fine?

For the rest, I think we would need to abstract and replace the NewCirros in the tests. In the past, we tried to avoid it, but in this case, it might be necessary. @akalenyu @EdDev wdyt?

vamsikrishna-siddu commented:
Hi @alicefr  I have Addressed your review comments. Can you please review this PR again Thanks

alicefr commented on tests/libvmifact/factory.go:
```diff
@@ -98,19 +106,28 @@ func NewAlpineWithTestTooling(opts ...libvmi.Option) *kvirtv1.VirtualMachineInst
 	}
+	if IsS390X() {
+		// Using Alpine for the s390x Architecture.
+		const s390xMinimalBootableMemory = "128Mi"
+		return s390xMinimalBootableMemory
```
```suggestion
		return  "128Mi"
```

vamsikrishna-siddu replied:
done

alicefr commented on tests/network/vmi_lifecycle.go:
```diff
@@ -59,7 +59,6 @@ var _ = SIGDescribe("[crit:high][arm64][vendor:cnv-qe@redhat.com][level:componen
 		Context("when virt-handler is responsive", func() {
 			DescribeTable("[Serial]VMIs shouldn't fail after the kubelet restarts", func(bridgeNetworking bool) {
 				var vmiOptions []libvmi.Option
-
```
nit:removed empty line

vamsikrishna-siddu replied:
added the empty line back.

alicefr commented on tests/vmi_iothreads_test.go:
```diff
@@ -216,7 +216,7 @@ var _ = Describe("[sig-compute]IOThreads", decorators.SigCompute, func() {
 			// there's only 6 threads expected because there's 6 total disks, even
 			// though the limit would have supported 8.
-			Entry("[test_id:3096]for four CPUs", 4, 6),
+			Entry("[test_id:3096]for four CPUs", 4, 7),
```
Why do we need to change from 6 to 7?

vamsikrishna-siddu replied:
Same reason as this comment.
https://github.com/kubevirt/kubevirt/pull/12516/files/58a731f3dc057f525d5d86e907c4d3f60d74e03c#r1734287174

alicefr commented on tests/vmi_lifecycle_test.go:
```diff
@@ -220,19 +221,24 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
-
+			vmi := libvmifact.NewGuestless(options...)
 			vmi = libvmops.RunVMIAndExpectLaunch(vmi, startupTimeout)
-
```
nit: removed empty line

vamsikrishna-siddu replied:
added the empty line back.

alicefr commented on tests/vmi_lifecycle_test.go:
```diff
@@ -287,11 +293,11 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 			Expect(reviewResponse.Details.Causes).To(HaveLen(2), "There should be 2 thing wrong in response")
-			Expect(reviewResponse.Details.Causes[0].Field).To(Equal("spec.domain.devices.disks[1].name"))
-			Expect(reviewResponse.Details.Causes[1].Field).To(Equal("spec.domain.devices.disks[2].name"))
+			Expect(reviewResponse.Details.Causes[0].Field).To(Equal("spec.domain.devices.disks[2].name"))
+			Expect(reviewResponse.Details.Causes[1].Field).To(Equal("spec.domain.devices.disks[3].name"))
```
Why do we need to change the index of the disks?

vamsikrishna-siddu replied:
Same reason as this comment.
https://github.com/kubevirt/kubevirt/pull/12516/files/58a731f3dc057f525d5d86e907c4d3f60d74e03c#r1734287174

alicefr commented on tests/vmipreset_test.go:
```diff
@@ -137,7 +137,7 @@ var _ = Describe("[rfe_id:609][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 			Expect(reviewResponse.Details.Causes).To(HaveLen(1))
-			Expect(reviewResponse.Details.Causes[0].Field).To(Equal("spec.domain.devices.disks[1]"))
+			Expect(reviewResponse.Details.Causes[0].Field).To(Equal("spec.domain.devices.disks[2]"))
```
Why do we need to change the disk index?

vamsikrishna-siddu replied:
Hi @alicefr  it is because of this change. https://github.com/kubevirt/kubevirt/pull/12516/files#diff-1ffe3878238ae2b0809d647ed00f3f2db70c6ff5d96b181dda8c08f907566949R83-R86

The Alpine already has two disks one is disk0 and cloudinit disk which is added during the creation due to the above change.  So, the test disk we are adding to the preset is at the location disks[2]

alicefr commented:
@vamsikrishna-siddu please refactor the commits and integrate the latest with the first. Otherwise, it becomes hard to review it since you changed stuff that was added in the first commit. Please, try to have a clean git history

alicefr commented:
+1 for the Guestless function, that looks much more better

alicefr commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -51,6 +55,29 @@ var (
+	loginToVMI = console.LoginToCirros
+
+	return vmi, loginToVMI, isAlpine
+
```
nit: additional empty line

vamsikrishna-siddu replied:
removed the additional empty line.

alicefr commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -51,6 +55,29 @@ var (
 )
 
+// GetVMGuestByArchitecture returns the VMGuest based on the Architecture.
+func GetVMGuestByArchitecture(opts ...libvmi.Option) (*v1.VirtualMachineInstance,
```
@EdDev @xpivarc what do you think about this function? I don't like the 3 return parameter much, but I don't see right now a better way

vamsikrishna-siddu commented:
> @vamsikrishna-siddu please refactor the commits and integrate the latest with the first. Otherwise, it becomes hard to review it since you changed stuff that was added in the first commit. Please, try to have a clean git history

Hi @alicefr  I have refactored the commits and integrated the latest commit with the first one. Can you please review once again.?

alicefr commented:
@vamsikrishna-siddu for me it looks good enough, thanks! I'd like the @EdDev and @xpivarc 's opinion on https://github.com/kubevirt/kubevirt/pull/12516#discussion_r1734252271. IMO, it could also be address in a follow up PR.

Can you please fix the lint job?

/approve
/hold
To let the other have time to review

lyarwood commented:
/cc

lyarwood commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -51,6 +55,28 @@ var (
+	var isAlpine bool
+	arch := putil.TranslateBuildArch()
+	// Cirros is not supported for s390x.
+	// Use Alpine instead of Cirros for s390x.
```
If this is the case shouldn't we just deprecate the use of CirrOS in our tests and move everything across to Alpine?

vamsikrishna-siddu replied:
Dropping cirros in favour of alpine would simplify our enabling e2e test for s390x, but i am not sure if there would be consequences for the other arches.
Might be a good discussion topic on the mailing list/next community meeting.

vamsikrishna-siddu replied:
I can do this way. But it might cause confusion. Because when someone reads the code, and they came across NewCirros in the test they will think that it will create a Cirros VMI. But when they run the test on the s390x cluster. It will create an Alpine VMI.

lyarwood replied:
Yup fair points. Can you still push ahead with the discussion around the removal of cirros on the ML?

lyarwood replied:
Discussing this offline it might even be better to skip the offending `CirrOS` tests for now, avoiding the need for this code and port them over to `Alpine` to allow them to run on `s390x` eventually.

vamsikrishna-siddu commented on tests/libvmifact/factory.go:
```diff
@@ -98,19 +106,27 @@ func NewAlpineWithTestTooling(opts ...libvmi.Option) *kvirtv1.VirtualMachineInst
+	arch := util.TranslateBuildArch()
+	// Use Alpine if the architecture is s390x.
+	if checks.IsS390X(arch) {
+		vmi := NewAlpine(opts...)
```
done

vamsikrishna-siddu commented on tests/libvmifact/factory.go:
```diff
@@ -98,19 +106,27 @@ func NewAlpineWithTestTooling(opts ...libvmi.Option) *kvirtv1.VirtualMachineInst
 
 func NewGuestless(opts ...libvmi.Option) *kvirtv1.VirtualMachineInstance {
-	opts = append(
-		[]libvmi.Option{libvmi.WithResourceMemory(qemuMinimumMemory())},
```
So, whenever we want to pass a different memory to the **`NewGuestless`** function. The above code is overriding it with the **`qemuMinimumMemory`** function. So, I have removed it from inside the function and passing it as a function argument.  This way we can avoid this issue, and another thing is that we don't need to explicitly call libvmi.New() in the tests. We can directly call the NewGuestless with the required options and it will take care. 
See here:
https://github.com/kubevirt/kubevirt/pull/12516/files#diff-d9842d3f25d0c190c12e603a60b0af75d5f52f6704866e9db63aeba68bd414f6L1227-R1251

lyarwood replied:
Can you point to an example where we use `NewGuestless` with something other than 
 `libvmi.WithResourceMemory(libvmifact.QemuMinimumMemory())`? The above link isn't working for me for some reason. Again given your changes to `QemuMinimumMemory` this seems pointless.

vamsikrishna-siddu replied:
can you check this once?
https://github.com/kubevirt/kubevirt/blob/6ed1da062e8765b857fe434fd74bedd81f8d3a64/tests/vmi_lifecycle_test.go#L240

lyarwood replied:
Ah sorry github hid the diff for that file!

vamsikrishna-siddu commented:
Hi @lyarwood 
I have done as you suggested here
https://github.com/kubevirt/kubevirt/pull/12516#discussion_r1743654856 and also   addressed your other review comments. Can you please review it once again.

vamsikrishna-siddu commented on tests/vmi_lifecycle_test.go:
```diff
@@ -220,17 +221,23 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 			}
-			vmi := libvmi.New(options...)
-
+			vmi := libvmifact.NewGuestless(options...)
```
Hi @lyarwood , Apoligies for unclear explanation from my side. Let me explain here clearly.
The libvmi.New function is not working for s390x. The VMI is not coming to running state. So, I am using Alpine for s390x.  So, I have moved the Alpine logic and libvmi.new(options) logic inside the NewGuestless function. 

```
func NewGuestless(opts ...libvmi.Option) *kvirtv1.VirtualMachineInstance {
	arch := util.TranslateBuildArch()
	// Use Alpine if the architecture is s390x.
	if checks.IsS390X(arch) {
		return NewAlpine(opts...)
	}
	return libvmi.New(opts...)
}
```
So, if we consider the above test. It requires 32Mi for the test to pass. If I directly call the libvmifact.NewGuestless without removing this part from NewGuestless function,
```
 opts = append(
		[]libvmi.Option{libvmi.WithResourceMemory(qemuMinimumMemory())},
```
 it will override the 32Mi which Iam passing in the function with QemuMinMemory i.e 1Mi and the test will fail.  So, I have removed this part of the code and explicitly passing the QemuMinMemory as a function argument.  So now the NewGuestless function will work as follows

```
NewGuestless(QemuMinMemory) --> it will work as the earlier  NewGuestless function 
NewGuestless(with our own memory and options). --> it will work as libvmi.New(options) if not s390x
NewGuestless(options) for s390x --> it will work as NewAlpine(options)
```

if i dont follow this way. I have to explicitly check for the s390x arch and call the NewAlpine fn.

```
if checks.IsS390x(arch) {
 vmi := libvmi.NewAlpine(options...)
}else{
  vmi := libvmi.New(options...)
}
```
I did this way because of this suggestion from one of the reviewer.
https://github.com/kubevirt/kubevirt/pull/12516#issuecomment-2309545558

lyarwood replied:
Thanks for taking the time to write that up but I'm still not convinced you need to change `NewGuestless` at all.

If `New` isn't working in this context for `s390x` then surely `New` or the test cases calling it that don't work on `s390x` should be changed.

That avoids the need to change every caller to `NewGuestless` and the very meaning of this helper (emphasis on it being _guestless_) as it's basically the same as `New` outside of `s390x` use with your changes.

vamsikrishna-siddu replied:
Hi @lyarwood ,  I have reverted the New Guestless function, and the files changes due to this to original.   I have also changed the test cases to reflect the changes.  Can you please review it again ?

vamsikrishna-siddu commented:
/retest

lyarwood commented on tests/network/vmi_lifecycle.go:
```diff
@@ -122,12 +123,17 @@ var _ = SIGDescribe("[crit:high][arm64][vendor:cnv-qe@redhat.com][level:componen
 				verifyDummyNicForBridgeNetwork(bridgeVMI)
 
 				vmIP := libnet.GetVmiPrimaryIPByFamily(bridgeVMI, k8sv1.IPv4Protocol)
 				dadCommand := fmt.Sprintf("sudo /usr/sbin/arping -D -I eth0 -c 2 %s | grep Received | cut -d ' ' -f 2\n", vmIP)
```
nit - would it not be easier to define this once without `sudo` and then add it back in if `!isAlpine`?

lyarwood commented on tests/vnc_test.go:
```diff
@@ -53,22 +53,28 @@ import (
 			vmi = libvmifact.NewGuestless()
+			arch := util.TranslateBuildArch()
+			if checks.IsS390X(arch) {
+				vmi = libvmifact.NewAlpine()
```
Why do these specific tests need to be Alpine?

IOW what is this VMI providing on s390x that the guestless VMI can't?

The answer to these questions should be documented as code comments here IMHO.

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -664,6 +677,7 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 				// Schedule a vmi and make sure that virt-handler gets evicted from the node where the vmi was started
 				// Note: we want VMI without any container
 				vmi = libvmifact.NewGuestless(libvmi.WithLogSerialConsole(false))
+
```
nit - whitespace

vamsikrishna-siddu replied:
done

lyarwood commented on tests/network/vmi_lifecycle.go:
```diff
@@ -48,7 +48,7 @@ import (
 )
 
-var _ = SIGDescribe("[crit:high][arm64][vendor:cnv-qe@redhat.com][level:component]", func() {
+var _ = SIGDescribe("[crit:high][arm64][vendor:cnv-qe@redhat.com][level:component]", decorators.WgS390x, func() {
```
nit - do you also need to add [s390x] to the description here? That said I'm slightly confused why we are hard coding arches into these strings If I'm honest.

lyarwood commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -91,8 +117,14 @@ func AdjustKubeVirtResource() {
 			},
 		},
 	}
+	arch := putil.TranslateBuildArch()
```
nit - inline below

vamsikrishna-siddu replied:
done

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -220,8 +221,15 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
-			options := []libvmi.Option{libvmi.WithResourceMemory("32Mi")}
+		DescribeTable("log libvirtd debug logs should be", decorators.WgS390x, func(vmiLabels, vmiAnnotations map[string]string, expectDebugLogs bool) {
+			arch := util.TranslateBuildArch()
+			var memory string
```
supernit 
```
memory := "32Mi"
if arch == "s390x" {
    memory = "128Mi"
}
```

vamsikrishna-siddu replied:
comment not applicable after reverting the changes.

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -231,6 +239,9 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 			vmi := libvmi.New(options...)
 
+			if arch == "s390x" {
+				vmi = libvmifact.NewAlpine(options...)
```
Again can we document why this required for this test in code comments.

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1212,8 +1225,19 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+		Context("with non default namespace", decorators.WgS390x, func() {
 			DescribeTable("[rfe_id:273][crit:high][vendor:cnv-qe@redhat.com][level:component]should log libvirt start and stop lifecycle events of the domain", func(alternativeNamespace bool) {
+				arch := util.TranslateBuildArch()
+				var memory string
```
supernit - as above, set defaults and drop the else.

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

lyarwood commented on tests/vnc_test.go:
```diff
@@ -346,8 +352,16 @@ func (h ResolutionMatcher) NegatedFailureMessage(actual interface{}) (message st
 	if videoType == "virtio" {
-		X = 1280
-		Y = 800
+		arch := util.TranslateBuildArch()
```
nit - inline

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

lyarwood commented on tests/vnc_test.go:
```diff
@@ -346,8 +352,16 @@ func (h ResolutionMatcher) NegatedFailureMessage(actual interface{}) (message st
-		Y = 800
+		arch := util.TranslateBuildArch()
+		// Use different resolution for s390x.
+		if checks.IsS390X(arch) {
```
supernit - same as above, define the default and drop the `else`.

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

lyarwood commented on tests/vnc_test.go:
```diff
@@ -346,8 +352,16 @@ func (h ResolutionMatcher) NegatedFailureMessage(actual interface{}) (message st
-		X = 1280
-		Y = 800
+		arch := util.TranslateBuildArch()
+		// Use different resolution for s390x.
```
Specifically because the VGA device provided by QEMU on s390x only supports this resolution? Can you document the answer here please.

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1224,11 +1248,11 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
-				)
+				vmi = libvmi.New(options...)
+				if arch == "s390x" {
+					vmi = libvmifact.NewAlpine(options...)
```
Again lets document the justification for this

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1212,8 +1225,19 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
-		Context("with non default namespace", func() {
+		Context("with non default namespace", decorators.WgS390x, func() {
 			DescribeTable("[rfe_id:273][crit:high][vendor:cnv-qe@redhat.com][level:component]should log libvirt start and stop lifecycle events of the domain", func(alternativeNamespace bool) {
+				arch := util.TranslateBuildArch()
```
nit - inline below

vamsikrishna-siddu replied:
review comment not applicable after reverting the changes.

vamsikrishna-siddu commented:
Hi @lyarwood , As Discussed, I am only running the tests which uses alpine VMI's, and I have reverted the changes for the tests that uses Cirros and Guestless.  I have also addressed your other review comments.  Can you please review it again. Thanks

lyarwood commented on tests/vmi_lifecycle_test.go:
```diff
@@ -1267,12 +1267,12 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
-						1. virt-controller kicks in, registers a deletion request on the launcher pod and K8s deletes the pod
-					       before virt-handler had a chance to set or check the deletion timestamp on the domain.
-						2. virt-handler detects the deletion timestamp on the domain and removes it.
+						        Since we deleted the VMI object, there are two possible outcomes and both are expected:
```
Accidental changes?

vamsikrishna-siddu replied:
yes, let me clean up them

vamsikrishna-siddu replied:
For some reason, My Visual studio code is formatting it this way when Iam saving the file. I am currently searching for a way to switch off the formatting.

vamsikrishna-siddu replied:
Done

lyarwood reviewed (changes requested):
Thanks this looks better now, looks like some accidental changes to clean up but otherwise this should be fine and we can follow up in new PRs for the remaining tests.

vamsikrishna-siddu commented:
/test pull-kubevirt-unit-test

xpivarc commented on pkg/util/util.go:
```diff
@@ -293,3 +293,24 @@ func GenerateKubeVirtGroupVersionKind(obj runtime.Object) (runtime.Object, error
+	case2:
+	  build_arch is s390x, which will be translated to s390x arch
+*/
+func TranslateBuildArch() string {
```
Do we expect to have this env during execution of the test?

vamsikrishna-siddu replied:
Yes

xpivarc replied:
We run the binary directly, what will set the env?

vamsikrishna-siddu replied:
This env will be set in the prow job that runs the e2e test for s390x.  We will set this env to `BUILD_ARCH="crossbuild-s390x"`

vamsikrishna-siddu commented:
https://github.com/kubevirt/kubevirt/pull/12516#discussion_r1749915935
If we look at here.  https://github.com/kubevirt/kubevirt/blob/main/tests/testsuite/kubevirtresource.go#L141 Whenever the CPUManager feature is enabled, it is looking for the CPUManager label on the worker nodes and if it is not able to find the label then it will not run the tests and it will stop the test execution saying label not found. The CPUManager is not supported for s390x. So, when this feature is enabled, it is not able to find the label on the s390x worker nodes and the tests are not running. So, I am only adding this feature in the default kubevirt configuration if the architecture is not s390x.

xpivarc commented:
> [#12516 (comment)](https://github.com/kubevirt/kubevirt/pull/12516#discussion_r1749915935) If we look at here. https://github.com/kubevirt/kubevirt/blob/main/tests/testsuite/kubevirtresource.go#L141 Whenever the CPUManager feature is enabled, it is looking for the CPUManager label on the worker nodes and if it is not able to find the label then it will not run the tests and it will stop the test execution saying label not found. The CPUManager is not supported for s390x. So, when this feature is enabled, it is not able to find the label on the s390x worker nodes and the tests are not running. So, I am only adding this feature in the default kubevirt configuration if the architecture is not s390x.

Thank you for the explanation, this is unfortunate but let's solve it another time.

alicefr commented:
@vamsikrishna-siddu how did we solve the cirros tests? Have you decided only to run the ones using Alpine?

vamsikrishna-siddu commented:
> @vamsikrishna-siddu how did we solve the cirros tests? Have you decided only to run the ones using Alpine?

Hi @alicefr  as Discussed here with @lyarwood ,  https://github.com/kubevirt/kubevirt/pull/12516#discussion_r1746918761 .I will start the initial e2e s390x tests by running only Alpine tests. Later I will start a Discussion in the kubevirt MailingList/ Community meeting for the Deprecation of use of Cirros in the tests . If community agrees, I will later enable the other tests by replacing the Cirros with alpine.

alicefr commented:
/unhold

vamsikrishna-siddu commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-compute
